### PR TITLE
Replaced open/close for some file handles with a context manager.

### PR DESCRIPTION
**Pull request by atombrella**
##### SUMMARY

I noticed a lot of these. I believe that it's more Pythonic to use `with ...: do stuff`, instead of explicitly opening and closing the fie handle. There are also many files that are not closed.

It's by no means complete, as I'm not sure if it'll be accepted.

I picked `Bugfix` as the issue type, as the three other options did not really match. It's style fixes, though.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
It's a generic style problem across many different components.

##### ADDITIONAL INFORMATION
https://www.youtube.com/watch?v=wf-BqAjZb8M has an example with `connect`/`disconnect` that is comparable.
```

### CODE REVIEW

**Comment by atombrella** [comment]
@bcoca It was made available as of 2.6, and 2.5 had `from __future__ import with_statement`. There are some more details in https://www.python.org/dev/peps/pep-0343/ So, support is no longer an obstacle :) Hopefully, Python 2.6 support will also be dropped at some point.

**Comment by atombrella** [comment]
And there are already quite a few context managers in the project already. `grep --color=always -E -r "^ +with " | less -R` gives you a clue :)

**Comment by Fred-sun** [comment]
@atombrella Thanks for your contribution this PR, Please fixes the CI fail , when you finished change, I will push for merge! Thank you very much!

**Comment by atombrella** [comment]
`ready_for_review`. I have appended a lot since I initially opened it, but will stop doing that.

Context managers can be used in other places, too.

@Fred-sun Could you take a look at relevant parts? Thank you.

File: docs/docsite/rst/dev_guide/developing_python_3.rst [hunk 1]
```diff
@@ -167,11 +167,11 @@ works on both versions:
     from ansible.module_utils._text import to_bytes
 
     filename = u'/var/tmp/くらとみ.txt'
-    f = open(to_bytes(filename), 'wb')
-    mtime = os.path.getmtime(to_bytes(filename))
-    b_filename = os.path.expandvars(to_bytes(filename))
-    if os.path.exists(to_bytes(filename)):
-        pass
+    with open(to_bytes(filename), 'wb') as f:
```
Comment (felixfontein) [inline docs/docsite/rst/dev_guide/developing_python_3.rst:170]:
```suggestion
    with open(to_bytes(filename), 'wb') as f:
        pass
```
I would de-dent the other lines. They are not supposed to use `f`; these lines are only a collection of things doing something with filenames.

File: lib/ansible/inventory/manager.py [hunk 2]
```diff
@@ -619,9 +619,8 @@ def subset(self, subset_pattern):
             # allow Unix style @filename data
             for x in subset_patterns:
                 if x[0] == "@":
-                    fd = open(x[1:])
-                    results.extend([to_text(l.strip()) for l in fd.read().split("\n")])
-                    fd.close()
+                    with open(x[1:]) as fd:
+                        results.extend([to_text(l.strip()) for l in fd.readlines()])
```
Comment (felixfontein) [inline lib/ansible/inventory/manager.py:623]: This change changes semantics: `fd.read().split('\n')` does not return the newlines, while `fd.readlines()` does.

I would simply do
```suggestion
                        results.extend([to_text(l.strip()) for l in fd.read().split("\n")])
```
instead.

File: lib/ansible/module_utils/known_hosts.py [hunk 3]
```diff
@@ -115,15 +115,13 @@ def not_in_host_file(self, host):
             continue
 
         try:
-            host_fh = open(hf)
+            with open(hf) as host_fh:
+                data = host_fh.readlines()
```
Comment (felixfontein) [inline lib/ansible/module_utils/known_hosts.py:119]: Same here. You are changing behavior.
```suggestion
                data = host_fh.read()
```

File: lib/ansible/module_utils/known_hosts.py [hunk 4]
```diff
@@ -115,15 +115,13 @@ def not_in_host_file(self, host):
...
         except IOError:
             hfiles_not_found += 1
             continue
-        else:
-            data = host_fh.read()
-            host_fh.close()
 
-        for line in data.split("\n"):
+        for line in data:
```
Comment (felixfontein) [inline lib/ansible/module_utils/known_hosts.py:124]:
```suggestion
        for line in data.split("\n"):
```

File: lib/ansible/modules/network/cnos/cnos_backup.py [hunk 5]
```diff
@@ -259,9 +259,8 @@ def main():
     # cnos.debugOutput(path[0])
     if not os.path.exists(path[0]):
         os.makedirs(path[0])
-    file = open(outputfile, "a")
-    file.write(output)
-    file.close()
+    with open(outputfile, "a") as fh:
```
Comment (felixfontein) [inline lib/ansible/modules/network/cnos/cnos_backup.py:262]:
```suggestion
    with open(outputfile, "a") as file:
```
It is a less intrusive change if you keep the original handle's name.

File: lib/ansible/modules/network/cnos/cnos_reload.py [hunk 6]
```diff
@@ -97,9 +97,8 @@ def main():
     output = output + str(cnos.run_cnos_commands(module, cmd))
 
     # Save it into the file
-    file = open(outputfile, "a")
-    file.write(output)
-    file.close()
+    with open(outputfile, "a") as fh:
```
Comment (felixfontein) [inline lib/ansible/modules/network/cnos/cnos_reload.py:100]: Same here.

File: lib/ansible/modules/system/hostname.py [hunk 7]
```diff
@@ -284,14 +273,14 @@ class RedHatStrategy(GenericStrategy):
...
-                        k, v = line.split('=')
-                        return v.strip()
-            finally:
-                f.close()
+                with open(self.NETWORK_FILE, 'rb') as f:
+                    for line in f.readlines():
+                        if line.startswith('HOSTNAME'):
+                            k, v = line.split('=')
+                            return v.strip()
+            except IOError:
```
Comment (felixfontein) [inline lib/ansible/modules/system/hostname.py:282]: Why are you ignoring `IOError`s? The original code did not do that!

File: lib/ansible/modules/system/hostname.py [hunk 8]
```diff
@@ -300,23 +289,23 @@ def set_permanent_hostname(self, name):
...
-            finally:
-                f.close()
+                with open(self.HOSTNAME_FILE, 'w+') as f:
+                    for line in f.readlines():
+                        if line.startswith('HOSTNAME'):
+                            lines.append("HOSTNAME=%s\n" % name)
+                            found = True
+                        else:
+                            lines.append(line)
+            except IOError as e:
```
Comment (felixfontein) [inline lib/ansible/modules/system/hostname.py:300]: Same here: why did you add this?

File: lib/ansible/modules/system/hostname.py [hunk 9]
```diff
@@ -300,23 +289,23 @@ def set_permanent_hostname(self, name):
...
             if not found:
                 lines.append("HOSTNAME=%s\n" % name)
-            f = open(self.NETWORK_FILE, 'w+')
             try:
-                f.writelines(lines)
-            finally:
-                f.close()
+                with open(self.NETWORK_FILE, 'w+') as f:
+                    f.writelines(lines)
+            except IOError as e:
```
Comment (felixfontein) [inline lib/ansible/modules/system/hostname.py:307]: Same here!

File: lib/ansible/modules/system/hostname.py [hunk 10]
```diff
@@ -537,45 +508,39 @@ def get_permanent_hostname(self):
         name = 'UNKNOWN'
         if not os.path.isfile(self.HOSTNAME_FILE):
             try:
-                open(self.HOSTNAME_FILE, "a").write("hostname=temporarystub\n")
+                with open(self.HOSTNAME_FILE, "a") as f:
+                    f.write("hostname=temporarystub\n")
             except IOError as e:
                 self.module.fail_json(msg="failed to write file: %s" %
                                           to_native(e), exception=traceback.format_exc())
-        try:
```
Comment (felixfontein) [inline lib/ansible/modules/system/hostname.py:544]: You are heavily changing semantics here! Previously, the next lines were executed always. Now, they are only executed if `not os.path.isfile(self.HOSTNAME_FILE)` is `True`!

File: lib/ansible/modules/system/known_hosts.py [hunk 11]
```diff
@@ -306,13 +306,12 @@ def compute_diff(path, found_line, replace_or_add, state, key):
         'after': '',
     }
     try:
-        inf = open(path, "r")
+        with open(path, "r") as inf:
+            diff['before'] = inf.read()
     except IOError as e:
```
Comment (felixfontein) [inline lib/ansible/modules/system/known_hosts.py:311]: You are slightly changing semantics here, but I think this is ok.

Before, `IOError`s were only caught when they were raised by `open()`. Now they are also caught when raised by `read()` or `close()`.

File: lib/ansible/modules/system/locale_gen.py [hunk 12]
```diff
@@ -103,15 +102,16 @@ def fix_case(name):
 def replace_line(existing_line, new_line):
     """Replaces lines in /etc/locale.gen"""
     try:
-        f = open("/etc/locale.gen", "r")
-        lines = [line.replace(existing_line, new_line) for line in f]
-    finally:
-        f.close()
+        with open("/etc/locale.gen", "r") as f:
+            lines = [line.replace(existing_line, new_line) for line in f.readlines()]
```
Comment (felixfontein) [inline lib/ansible/modules/system/locale_gen.py:106]:
```suggestion
            lines = [line.replace(existing_line, new_line) for line in f]
```
Matches the original code more closely.

File: lib/ansible/modules/system/locale_gen.py [hunk 13]
```diff
@@ -103,15 +102,16 @@ def fix_case(name):
...
+    except IOError:
```
Comment (felixfontein) [inline lib/ansible/modules/system/locale_gen.py:107]: This is changing semantics!

File: lib/ansible/modules/system/locale_gen.py [hunk 14]
```diff
@@ -103,15 +102,16 @@ def fix_case(name):
...
+        pass
+
     try:
-        f = open("/etc/locale.gen", "w")
-        f.write("".join(lines))
-    finally:
-        f.close()
+        with open("/etc/locale.gen", "w") as f:
+            f.write("".join(lines))
+    except IOError:
```
Comment (felixfontein) [inline lib/ansible/modules/system/locale_gen.py:113]: Same here. Don't do that!

File: lib/ansible/modules/system/locale_gen.py [hunk 15]
```diff
@@ -122,15 +122,13 @@ def set_locale(name, enabled=True):
...
-        f = open("/etc/locale.gen", "w")
-        f.write("".join(lines))
-    finally:
-        f.close()
+        with open("/etc/locale.gen", "r") as f:
+            lines = [re.sub(search_string, new_string, line) for line in f]
+
+        with open("/etc/locale.gen", "w") as f:
+            f.write("".join(lines))
+    except IOError:
```
Comment (felixfontein) [inline lib/ansible/modules/system/locale_gen.py:130]: Same here. Why?

File: lib/ansible/modules/system/mount.py [hunk 16]
```diff
@@ -521,14 +521,8 @@ def get_linux_mounts(module, mntinfo_file="/proc/self/mountinfo"):
...
-        f = open(mntinfo_file)
-    except IOError:
-        return
-
-    lines = map(str.strip, f.readlines())
-
-    try:
-        f.close()
+        with open(mntinfo_file) as f:
+            lines = map(str.strip, f.readlines())
```
Comment (felixfontein) [inline lib/ansible/modules/system/mount.py:525]: You are changing semantics here. Originally, errors while opening the file were ignored, the function simply returned. Errors while reading the file's content were escalated, and errors while closing the file were handled (by failing the module).

Now, you're handling **all** errors by failing the module. You should definitely **not** handle file not found errors this way! (No idea about other errors while opening the file.)

File: lib/ansible/modules/system/mount.py [hunk 17]
```diff
@@ -521,14 +521,8 @@ def get_linux_mounts(module, mntinfo_file="/proc/self/mountinfo"):
...
     except IOError:
         module.fail_json(msg="Cannot close file %s" % mntinfo_file)
```
Comment (felixfontein) [inline lib/ansible/modules/system/mount.py:527]: Also, this error message is now wrong.

File: lib/ansible/modules/system/service_facts.py [hunk 18]
```diff
@@ -184,13 +184,13 @@ class SystemctlScanService(BaseService):
     def systemd_enabled(self):
         # Check if init is the systemd command, using comm as cmdline could be symlink
         try:
-            f = open('/proc/1/comm', 'r')
+            with open('/proc/1/comm', 'r') as f:
```
Comment (felixfontein) [inline lib/ansible/modules/system/service_facts.py:187]: Same here: you now not only ignore errors during opening, but also during reading. This is potentially changing semantics!

**Review by felixfontein — CHANGES_REQUESTED** [review]
This fixes quite a lot of file handling bugs hidden in the Ansible codebase. Unfortunately, right now it also contains a few semantics changes which are not ok. Please check out my review comments.

I don't know network modules well. Do they also execute Python code on network devices (which could have a very old Python)? I guess someone from the networks team should definitely review this.

File: lib/ansible/modules/network/cnos/cnos_backup.py (hunk 5 above)
Comment (atombrella) [reply lib/ansible/modules/network/cnos/cnos_backup.py:262]: OK. Changed. I'm getting warnings that `file` shadows a built-in; I had in mind if I were changing the line anyway, changing the variable name wouldn't matter.

File: lib/ansible/modules/system/known_hosts.py (hunk 11 above)
Comment (atombrella) [reply lib/ansible/modules/system/known_hosts.py:311]: Yes. Agreed. It's a result of using context managers to handle files.

File: lib/ansible/modules/system/mount.py (hunk 16 above)
Comment (atombrella) [reply lib/ansible/modules/system/mount.py:525]: Yes. I don't really see how it could be handled differently, in a nice way. There are some error codes that may be useful, https://docs.python.org/2/library/errno.html The exception for opening the file can be caused by different events. Can we consider a compromise to simplify this so that the `IOError` is used for the different events? I think that's also cleaner and more maintainable. Python 3 has `FileNotFoundError` but it's not available in Python 2.

File: lib/ansible/modules/system/mount.py (hunk 17 above)
Comment (atombrella) [reply lib/ansible/modules/system/mount.py:527]: Yes. I did not notice that :( As written above, converting it to a context manager may cause some slight semantic changes. Python 2 makes it cumbersome to catch the file not found error (it requires errno and I'm not sure if it's enough to check `exception.errno == errno.ENOENT`). I think that some of the explicit open error checking and error with closing the file should be reconsidered, and if they can be merged.

Comment (felixfontein) [reply lib/ansible/modules/system/mount.py:527]: How about splitting up the PR into multiple ones? At least one for the cases where semantics have to change (probably best one PR per module, because it will need discussion for every one of them, and having one module per PR increases the chance that some of them will be merged, even if some others are not or take forever), and one for the "obvious" changes.

Comment (atombrella) [reply lib/ansible/modules/system/mount.py:527]: Yes. I think that'd be the obvious outcome of some of these discussions. Thanks for the suggestions :) Ansible seems to be somewhat conservative about changes, for good reason, considering that it's in use by quite many people, and in very diverse contexts.

**Comment by jborean93** [comment]
@atombrella it seems like you are creating smaller PRs to make these changes which is great. It sounds like we can close this bigger PR as tackling it in smaller chunks usually leads to easier reviews and faster merged. Are you ok with closing this one or are you still trying to get this through in this PR?

**Comment by atombrella** [comment]
> @atombrella it seems like you are creating smaller PRs to make these changes which is great. It sounds like we can close this bigger PR as tackling it in smaller chunks usually leads to easier reviews and faster merged. Are you ok with closing this one or are you still trying to get this through in this PR?

Yes. It should be closed. Thanks for reminding me.
